### ISSUE TITLE
Deserialisation of Object types by explicitly passing the type info

### ISSUE THREAD

**Issue by supercharger** [protostuff/protostuff#152]
I can deserialise following MessageTransfer type if the serialization data contains type metadata http://www.protostuff.io/documentation/runtime-schema/#objectschema-dynamic . It won't work if the serialization is done from another language using vanilla protobuff which is kind of obvious as type info is required.  In one use case, Application has the type info when it is deserializing ( a case where we call methods at runtime, so we know what type the params are) , But there is no way to pass that to protostuff.  Please see the following example.

```java
public class MessageTransfer {
        private int length;
        private Object requestParam1;
        private Object requestParam2;
       // left getters and setters for brevity.
}
```
And
```java

/**
 * Only works if InputStream is serialized by protostuff, not with vanilla protobuf
*/
public MessageTransfer deserialize(InputStream in, Class[] paramTypes) {
    Schema<Object> schema = (Schema<Object>) RuntimeSchema.getSchema(MessageTransfer.class);
     // But it won't work because MessageTransfer contains fields of type Object.
    
    // MessageTransfer object contains requestParam1 of Object type. This will only work if InputStream 
    // is contains the serialization data done by protostuff 
    //even though I do know what type requestParam1,requestParam2 are  there is no way I can pass that
}
```

***Things I have tried:***

* Protostuff internally doing ```java.lang.reflect.Field```, So I tried creating instance of MessageTransfer with actual types and use that to create instance. But, that failed as ```java.lang.reflect.Field``` is of compile time.

* Also, I tried creating  directly fields and then the RuntimeSchema type, but it failed again with same reason of Field.getType() returning compile type class.
```java
for(java.lang.reflect.Field refField : msgTransObj.getClass().getDeclaredFields()) {
       final Field<MessageTransfer> field = RuntimeFieldFactory.getFieldFactory(
                      classes[fieldMapping - 1], ID_STRATEGY).create(fieldMapping, refField.getName(), refField,ID_STRATEGY);
    fields.add(field);
    fieldMapping++;
}
messageTransSchema = new RuntimeSchema(msgTransObj.getClass(), fields, RuntimeEnv.newInstantiator(msgTransObj.getClass()));
```
* While protobuf generated code will be able to do this, but the major reason to use protostuff is to avoid that.

Basically, We need an API to pass the type info as an alternative to ObjectSchema because it won't work with vanilla protobuf.

PS: originally discussed in this [google group](https://groups.google.com/forum/#!topic/protostuff/tbw3TFwJhto).

**Comment by kshchepanovskyi** [comment]
@supercharger 

What about following:

```java
		@SuppressWarnings("unchecked")
		public static List<Object> deserialize(InputStream in, Class<?>[] paramTypes) {
			CodedInput codedInput = new CodedInput(in, false);
			List<Object> result = new ArrayList<Object>();
			for (Class<?> paramType : paramTypes) {
					Schema<Object> schema = (Schema<Object>) RuntimeSchema.getSchema(paramType);
					Object param = readParam(codedInput, schema);
					result.add(param);

			}
			return result;
		}

		private static Object readParam(CodedInput in, Schema<Object> schema)  {
			try {
				Object param = schema.newMessage();
				// check mergeObject implementation
				// first you should read message size
				// then message data
				in.mergeObject(param, schema);
				return param;
			} catch (IOException e) {
				throw new RuntimeException(e);
			}
		}
```

I did not test this code yet - just want to check if I understand you correctly.

**Comment by supercharger** [comment]
@kshchepanovskyi 
No, Top level object is fixed, 
```java
Schema<Object> schema = (Schema<Object>) RuntimeSchema.getSchema(MessageTransfer.class);
// But it won't work because MessageTransfer contains fields of type Object.
```
As said, it won't work because MessageTransfer contains fields of type Object. But, It works when we deserialize the message which is serialized by protostuff because of [objectschema-dynamic](http://www.protostuff.io/documentation/runtime-schema/#objectschema-dynamic)  but not the one when it is deserialized by vanilla protobuf. So natural question is, how can it deserialize when it doesn't type info, So need a way to explicitly pass the type info

**Comment by kshchepanovskyi** [comment]
@supercharger 

You can ignore `length` when you know exact types of your fields and you can parse them.
But it is not always like that: one of typical situations is when remote side does not know how to parse a part of your message (unknown field). So protobuf parser simply skips this part of the message (for performance reasons) and goes to next field.

[io.protostuff.CodedInput#skipField`](https://github.com/protostuff/protostuff/blob/45b620dfe482bb1a20640bd0a6dcfcbc128e74b5/protostuff-core/src/main/java/io/protostuff/CodedInput.java#L146-L146):

```java
 public boolean skipField(final int tag) throws IOException
    {
        switch (WireFormat.getTagWireType(tag))
        {
            case WireFormat.WIRETYPE_VARINT:
                readInt32();
                return true;
            case WireFormat.WIRETYPE_FIXED64:
                readRawLittleEndian64();
                return true;
            case WireFormat.WIRETYPE_LENGTH_DELIMITED:
                // here we use this number, so we do not need to check what is inside of unknown block
                skipRawBytes(readRawVarint32());
                return true;
```

**Comment by supercharger** [comment]
@kshchepanovskyi Thanks a lot for answering !!
